maas-api/internal/token/handler.go

```diff
@@ -29,40 +31,82 @@ func NewHandler(name string, manager TokenManager) *Handler {
 	}
 }
 
-// ExtractUserInfo validates kubernetes tokens.
-func (h *Handler) ExtractUserInfo(reviewer *Reviewer) gin.HandlerFunc {
+// parseGroupsHeader parses the X-MAAS-GROUP header which comes as a JSON array.
+// Format: "[\"group1\",\"group2\",\"group3\"]" (JSON-encoded array string).
+func parseGroupsHeader(header string) ([]string, error) {
+	if header == "" {
+		return nil, errors.New("header is empty")
+	}
+
+	// Try to unmarshal as JSON array directly
+	var groups []string
+	if err := json.Unmarshal([]byte(header), &groups); err != nil {
+		return nil, fmt.Errorf("failed to parse header as JSON array: %w", err)
+	}
+
+	if len(groups) == 0 {
+		return nil, errors.New("no groups found in header")
+	}
+
+	// Trim whitespace from each group
+	for i := range groups {
+		groups[i] = strings.TrimSpace(groups[i])
+	}
+
+	return groups, nil
+}
+
+// ExtractUserInfo extracts user information from X-MAAS-* headers set by the auth policy.
+func (h *Handler) ExtractUserInfo() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		authHeader := c.GetHeader("Authorization")
-		if authHeader == "" {
-			c.JSON(http.StatusUnauthorized, gin.H{"error": "Authorization header required"})
+		username := strings.TrimSpace(c.GetHeader("X-MAAS-USERNAME"))
+		groupHeader := c.GetHeader("X-MAAS-GROUP")
```

Comment:
Probably worth extracting to constants for consistency - we also use it in tests.